sygnal/webpushpushkin.py

```diff
@@ -0,0 +1,225 @@
+QUEUE_TIME_HISTOGRAM = Histogram(
+    "sygnal_webpush_queue_time",
+    "Time taken waiting for a connection to webpush endpoint",
+)
+
+SEND_TIME_HISTOGRAM = Histogram(
+    "sygnal_webpush_request_time", "Time taken to send HTTP request to webpush endpoint"
+)
+
+PENDING_REQUESTS_GAUGE = Gauge(
+    "sygnal_pending_webpush_requests",
+    "Number of webpush requests waiting for a connection",
+)
+
+ACTIVE_REQUESTS_GAUGE = Gauge(
+    "sygnal_active_webpush_requests", "Number of webpush requests in flight"
+)
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_MAX_CONNECTIONS = 20
+
+
+class WebpushPushkin(ConcurrencyLimitedPushkin):
+    """
+    Pushkin that relays notifications to Google/Firebase Cloud Messaging.
+    """
+
+    UNDERSTOOD_CONFIG_FIELDS = {
+        "type",
+        "max_connections",
+        "vapid_private_key",
+        "vapid_contact_email",
+    } | ConcurrencyLimitedPushkin.UNDERSTOOD_CONFIG_FIELDS
+
+    def __init__(self, name, sygnal, config):
+        super(WebpushPushkin, self).__init__(name, sygnal, config)
+
+        nonunderstood = self.cfg.keys() - self.UNDERSTOOD_CONFIG_FIELDS
+        if nonunderstood:
+            logger.warning(
+                "The following configuration fields are not understood: %s",
+                nonunderstood,
+            )
+
+        self.http_pool = HTTPConnectionPool(reactor=sygnal.reactor)
+        self.max_connections = self.get_config(
+            "max_connections", DEFAULT_MAX_CONNECTIONS
+        )
+        self.connection_semaphore = DeferredSemaphore(self.max_connections)
+        self.http_pool.maxPersistentPerHost = self.max_connections
+
+        tls_client_options_factory = ClientTLSOptionsFactory()
+
+        # use the Sygnal global proxy configuration
+        proxy_url = sygnal.config.get("proxy")
+
+        self.http_agent = ProxyAgent(
+            reactor=sygnal.reactor,
+            pool=self.http_pool,
+            contextFactory=tls_client_options_factory,
+            proxy_url_str=proxy_url,
+        )
+        self.http_agent_wrapper = HttpAgentWrapper(self.http_agent)
+
+        privkey_filename = self.get_config("vapid_private_key")
+        if not privkey_filename:
+            raise PushkinSetupException("'vapid_private_key' not set in config")
+        if not os.path.exists(privkey_filename):
+            raise PushkinSetupException("path in 'vapid_private_key' does not exist")
+        self.vapid_private_key = Vapid.from_file(private_key_file=privkey_filename)
+        vapid_contact_email = self.get_config("vapid_contact_email")
+        if not vapid_contact_email:
+            raise PushkinSetupException("'vapid_contact_email' not set in config")
+        self.vapid_claims = {"sub": "mailto:{}".format(vapid_contact_email)}
+
+    async def _dispatch_notification_unlimited(self, n, device, context):
+        p256dh = device.pushkey
+        if not isinstance(device.data, dict):
+            logger.info("device.data is not a dict, reject pushkey")
```

Comment:
This should probably be a warning (or maybe an error) log?

Probably the same for the other logging below.

Also should these log lines include some info about the device or something so it can be traced?